File: src/iit/core/system.py
from dataclasses import dataclass

import numpy as np

from src.iit.core.ncube import NCube


@dataclass(slots=True)
class System:
    estado_inicial: tuple[int, ...]
    ncubos: tuple[NCube, ...]

    @property
    def indices(self) -> tuple[int, ...]:
        return tuple(c.indice for c in self.ncubos)

    @property
    def dims(self) -> tuple[int, ...]:
        return self.ncubos[0].dims if self.ncubos else ()

    def condicionar(self, indices: tuple[int, ...]) -> "System":
        nuevos = tuple(
            c.condicionar(indices, self.estado_inicial)
            for c in self.ncubos
            if c.indice not in indices
        )
        return System(self.estado_inicial, nuevos)

    def substraer(
        self, alcance: tuple[int, ...], mecanismo: tuple[int, ...]
    ) -> "System":
        nuevos = tuple(
            c.marginalizar(mecanismo) for c in self.ncubos if c.indice not in alcance
        )
        return System(self.estado_inicial, nuevos)

    def bipartir(
        self, alcance: tuple[int, ...], mecanismo: tuple[int, ...]
    ) -> "System":
        nuevos = tuple(
            c.marginalizar(tuple(d for d in c.dims if d not in mecanismo))
            if c.indice in alcance
            else c.marginalizar(mecanismo)
            for c in self.ncubos
        )
        return System(self.estado_inicial, nuevos)

    def distribucion_marginal(self) -> np.ndarray:
        return np.array(
            [c[self.estado_inicial] for c in self.ncubos],
            dtype=np.float32,
        )

    def __str__(self) -> str:
        sub_dims = self.dims
        cubos_info = [f"{c}" for c in self.ncubos]
        return (
            f"\nSystem(indices={self.indices}, dims={sub_dims})"
            f"\nInitial state: {self.estado_inicial}"
            f"\nNCubes:\n" + "\n".join(cubos_info)
        )

File: iit/src/core/system.rs
// ============ SYSTEM ============

use std::fmt;

use super::ncube::NCube;

#[derive(Debug, Clone)]
pub struct System {
    pub estado_inicial: Vec<u8>,
    pub ncubos: Vec<NCube>,
}

impl System {
    pub fn new(estado_inicial: Vec<u8>, ncubos: Vec<NCube>) -> Self {
        Self {
            estado_inicial,
            ncubos,
        }
    }

    pub fn indices_ncubos(&self) -> Vec<u8> {
        self.ncubos.iter().map(|c| c.indice).collect()
    }

    pub fn dims_ncubos(&self) -> Vec<u8> {
        self.ncubos
            .first()
            .map(|c| c.dims.clone())
            .unwrap_or_default()
    }

    pub fn condicionar(&self, indices: &[u8]) -> Self {
        let nuevos: Vec<NCube> = self
            .ncubos
            .iter()
            .filter(|c| !indices.contains(&c.indice))
            .map(|c| c.condicionar(indices, &self.estado_inicial))
            .collect();

        Self::new(self.estado_inicial.clone(), nuevos)
    }

    pub fn substraer(&self, alcance_idx: &[u8], mecanismo_dims: &[u8]) -> Self {
        let nuevos: Vec<NCube> = self
            .ncubos
            .iter()
            .filter(|c| !alcance_idx.contains(&c.indice))
            .map(|c| c.marginalizar(mecanismo_dims))
            .collect();

        Self::new(self.estado_inicial.clone(), nuevos)
    }

    // pub fn bipartir(&self, alcance: &[u8], mecanismo: &[u8]) -> Self {
    //     let nuevos: Vec<NCube> = self
    //         .ncubos
    //         .iter()
    //         .map(|c| {
    //             if alcance.contains(&c.indice) {
    //                 let complement: Vec<u8> = c
    //                     .dims
    //                     .iter()
    //                     .filter(|d| !mecanismo.contains(d))
    //                     .copied()
    //                     .collect();
    //                 c.marginalizar(&complement)
    //             } else {
    //                 c.marginalizar(mecanismo)
    //             }
    //         })
    //         .collect();

    //     Self::new(self.estado_inicial.clone(), nuevos)
    // }

    pub fn distribucion_marginal(&self) -> Vec<f32> {
        self.ncubos
            .iter()
            .map(|c| c.value_at(&self.estado_inicial))
            .collect()
    }
}

impl fmt::Display for System {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(
            f,
            "System(indices={:?}, dims={:?})",
            self.indices_ncubos(),
            self.dims_ncubos()
        )?;
        writeln!(f, "  estado_inicial={:?}", self.estado_inicial)?;
        writeln!(f, "  NCubes:")?;
        for c in &self.ncubos {
            writeln!(f, "    {}", c)?;
        }
        Ok(())
    }
}
 